Document/0x05e-Testing-Cryptography.md

```diff
@@ -6,11 +6,26 @@
 
 A general rule in app development is that one should never attempt to invent their own cryptography. In mobile apps in particular, any form of crypto should be implemented using existing, robust implementations. In 99% of cases, this simply means using the data storage APIs and cryptographic libraries that come with the mobile OS.
 
-Android cryptography APIs are based on the Java Cryptography Architecture (JCA). JCA separates the interfaces and implementation, making it possible to include several cryptographic service providers that can implement sets of cryptographic algorithms. Most of the JCA interfaces and classes are defined in the `java.security.*` and `javax.crypto.*` packages. In addition, there are Android specific packages `android.security.*` and `android.security.keystore.*`.
+Android cryptography APIs are based on the Java Cryptography Architecture (JCA). JCA separates the interfaces and implementation, making it possible to include several security providers <sup>[8]</sup> that can implement sets of cryptographic algorithms. Most of the JCA interfaces and classes are defined in the `java.security.*` and `javax.crypto.*` packages. In addition, there are Android specific packages `android.security.*` and `android.security.keystore.*`.
 
 The list of providers included in Android varies between versions of Android and the OEM-specific builds. Some provider implementations in older versions are now known to be less secure or vulnerable. Thus, Android applications should not only choose the correct algorithms and provide good configuration, in some cases they should also pay attention to the strength of the implementations in the legacy providers.
+You can list the set of existing providers as follows:
+
+```java
+StringBuilder builder = new StringBuilder();
+for (Provider provider : Security.getProviders()) {
+    builder.append(provider.getName())
+            .append(" ")
+            .append(provider.getVersion())
+            .append("(")
+            .append(provider.getInfo())
+            .append(")\n");
+}
+String providers = builder.toString();
+//now display the string on the screen or in the logs for debugging.
+```
```

Comment:
Might be useful to add a short, 1-3 line example of the output, along with the device/OS info, e.g., "Below is an excerpt of the output from Nexus 6P running Android 7.1.1, security patch level June 2017"